platform/api/packet/api.go

```diff
@@ -0,0 +1,603 @@
+	return fmt.Sprintf(`#!ipxe
+set base-url %s
+kernel ${base-url}/coreos_production_pxe.vmlinuz initrd=coreos_production_pxe_image.cpio.gz coreos.first_boot=1 coreos.config.url=%s console=%s
+initrd ${base-url}/coreos_production_pxe_image.cpio.gz
+boot`, a.opts.InstallerImageURL, userdataURL, linuxConsole[a.opts.Board])
+}
+
+// device creation seems a bit flaky, so try a few times
+func (a *API) createDevice(hostname, ipxeScriptURL string) (device *packngo.Device, err error) {
+	for tries := apiRetries; tries >= 0; tries-- {
+		var response *packngo.Response
+		device, response, err = a.c.Devices.Create(&packngo.DeviceCreateRequest{
+			ProjectID:     a.opts.Project,
+			Facility:      a.opts.Facility,
+			Plan:          a.opts.Plan,
+			BillingCycle:  "hourly",
+			HostName:      hostname,
+			OS:            "custom_ipxe",
+			IPXEScriptUrl: ipxeScriptURL,
+			Tags:          []string{"mantle"},
+		})
+		if err == nil || response.StatusCode != 500 {
+			return
+		}
+		if tries > 0 {
+			time.Sleep(apiRetryInterval)
+		}
+	}
+	return
+}
+
+func (a *API) startConsole(deviceID string, console Console) error {
+	ready := make(chan error)
+
+	runner := func() error {
+		defer console.Close()
+
+		client, err := console.SSHClient("sos."+a.opts.Facility+".packet.net", deviceID)
+		if err != nil {
+			return fmt.Errorf("couldn't create SSH client for %s console: %v", deviceID, err)
+		}
+		defer client.Close()
+
+		session, err := client.NewSession()
+		if err != nil {
+			return fmt.Errorf("couldn't create SSH session for %s console: %v", deviceID, err)
+		}
+		defer session.Close()
+
+		reader := newBlockingReader()
+		defer reader.Close()
+
+		session.Stdin = reader
+		session.Stdout = console
+		if err := session.Shell(); err != nil {
+			return fmt.Errorf("couldn't start shell for %s console: %v", deviceID, err)
+		}
+
+		// cause startConsole to return
+		ready <- nil
+
+		err = session.Wait()
+		_, ok := err.(*ssh.ExitMissingError)
+		if err != nil && !ok {
+			plog.Errorf("%s console session failed: %v", deviceID, err)
+		}
+		return nil
+	}
+	go func() {
+		err := runner()
+		if err != nil {
+			ready <- err
+		}
+	}()
+
+	return <-ready
+}
+
+func (a *API) waitForActive(deviceID string) (*packngo.Device, error) {
+	for tries := launchTimeout / launchPollInterval; tries >= 0; tries-- {
+		device, _, err := a.c.Devices.Get(deviceID)
+		if err != nil {
+			return nil, fmt.Errorf("querying device: %v", err)
+		}
+		if device.State == "active" {
+			return device, nil
+		}
+		if tries > 0 {
+			time.Sleep(launchPollInterval)
+		}
+	}
+	return nil, fmt.Errorf("timed out waiting for device")
+}
+
+// Connect to the discard port and wait for the connection to close,
+// indicating that install is complete.
+func waitForInstall(address string) (err error) {
+	deadline := time.Now().Add(installTimeout)
+	dialer := net.Dialer{
+		Timeout: installPollInterval,
+	}
+	for tries := installTimeout / installPollInterval; tries >= 0; tries-- {
```

Comment:
True, but it would make the `tries > 0` test at the bottom of the loop rather more awkward.